### PR TITLE
implemented pad function

### PR DESCRIPTION
**Pull request by Aman-Godara**
#394 

Task:

- [x] Implemented pad function
- [x] add test cases
- [x] document the function
- [x] Elemental or Pure? -> Pure

### CODE REVIEW

**Comment by Aman-Godara** [comment]
While I was writing code I got **ambiguous interface error** to which I preferred to keep `pad_with` argument as non-optional (to support backward compatibility) over other possible solutions.

**Comment by ivan-pi** [comment]
Nice work @Aman-Godara!

Some comments:
* What happens when the `pad_with` dummy variable is of `type(string_type)` and the length is larger than 1? Wouldn't it be easier to limit `pad_with` to `character(len=1)` for now?
* In most cases I imagine padding will be performed with the whitespace character `' '`. For such cases it is inconvenient to have to specify `pad_with = ' '`.

What I propose is to have the following versions (and analogous for `padr`):

```
    interface padl
        module procedure :: padl_string_default      ! whitespace
        module procedure :: padl_string_pad_with     ! caller provided padding symbol
        module procedure :: padl_char_default
        module procedure :: padl_char_pad_with
    end interface padl
```

If we decide to allow `len(pad_with) > 1` in the future, I believe it remains possible to do it in a backward compatible way for both string_type and character. Even if we were to allow `len(pad_with) > 1` I still believe most usage will be with character literals like `pad_with = 'aabbcc'`. The case where you would want to pass a variable of `type(string_type)` seems unlikely to me, but can be always added later if someone requests it.

So I guess my bottom line is that having `pad_with` of `type(string_type)` is not worth the implementation/maintenance effort at this early stage.

**Comment by Aman-Godara** [comment]
> What happens when the pad_with dummy variable is of type(string_type) and the length is larger than 1? Wouldn't it be easier to limit pad_with to character(len=1) for now?

Yes, `pad_with` is limited to character(len=1) as of now. If a user gives `pad_with` as `type(string_type)` with length > 1, function will only consider first character for padding.
Also, when user gives `""` (empty string) as `pad_with` padding is performed but with default `" "` (1 space).

I also liked the interface suggested by you let me make the changes.

***
Also, how about making the function **elemental**?
The only problem in doing so is that if a user gives an input like this:
`padl(["abcde", "bcd", "abc"], 6, "#")` output will be: 
`["#abcde", "#bcd  ", "#abc  "]` which actually looks as if the input got its right side padded with `" "` (1 space) and left with `"#"`.

Except this case in all other cases having an **elemental** `pad` function will be quite useful as length of outputs is any way going to be same for all the possible cases (including the one mentioned above) so there is no problem in returning the output.
OR
we can also keep `padl_char_pad_with` and `padl_char_default` as `pure` as of now, and make `padl_string_pad_with` and `padl_string_default` as **elemental** because they don't have any of these problems highlighted above and are quite useful if a user is interested in padding many strings at once.

**Comment by Aman-Godara** [comment]
I didn't consider the case where a user gives an array in `output_length` argument as such:
`pad_with(["abc", "bcd"], [5, 6], '#')`
where returning the output `["abc##", "bcd###"]` in an array won't be possible.
Thus, I am keeping all functions as `pure` for now.

File: src/stdlib_strings.f90 [hunk 1]
```diff
@@ -644,38 +654,48 @@ end function padl_char_char
...
-    pure function padr_char_string(string, output_length, pad_with) result(res)
+    pure function padr_char_default(string, output_length) result(res)
         character(len=*), intent(in) :: string
         integer, intent(in) :: output_length
-        type(string_type), intent(in) :: pad_with
         character(len=max(len(string), output_length)) :: res
 
-        res = padr_char_char(string, output_length, char(pad_with))
-    end function padr_char_string
+        res = padr_char_char(string, output_length, " ")
```
Comment (ivan-pi) [inline src/stdlib_strings.f90:685]: For the default it should be possible to do simply
```
res = string
```
getting rid of one function and branch (if) call.

But I would need to consult the Fortran standard to guarantee the padding symbol will be whitespace under all possible circumstances.

Comment (Aman-Godara) [reply src/stdlib_strings.f90:685]: on page 162 of https://j3-fortran.org/doc/year/18/18-007r1.pdf I have got this:
image

Comment (ivan-pi) [reply src/stdlib_strings.f90:685]: The second bullet point sounds like what we want. Thanks for finding and referencing this.

File: src/stdlib_strings.f90 [hunk 2]
```diff
@@ -79,6 +79,40 @@ module stdlib_strings
...
+        !module procedure :: replace_all_string_char_string
+        !module procedure :: replace_all_char_string_string
+        !module procedure :: replace_all_string_char_char
+        !module procedure :: replace_all_char_string_char
+        !module procedure :: replace_all_char_char_string
+        module procedure :: replace_all_char_char_char
+    end interface replace_all
+
+    !> Left pad the input string
+    !> [Specifications](link to the specs - to be completed)
```
Comment (awvwgk) [inline src/stdlib_strings.f90:97]:
```suggestion
    !> [Specifications](../page/specs/stdlib_strings.html#padl)
```

File: src/stdlib_strings.f90 [hunk 3]
```diff
@@ -79,6 +79,40 @@ module stdlib_strings
...
+    !> Version: experimental
+    interface padl
+        module procedure :: padl_string_default
+        module procedure :: padl_string_pad_with
+        module procedure :: padl_char_default
+        module procedure :: padl_char_pad_with
+    end interface padl
+
+    !> Right pad the input string
+    !> [Specifications](link to the specs - to be completed)
```
Comment (awvwgk) [inline src/stdlib_strings.f90:107]:
```suggestion
    !> [Specifications](../page/specs/stdlib_strings.html#padr)
```

**Review by awvwgk — APPROVED** [review]
Only two minor comments, otherwise looks good to me.

File: src/stdlib_strings.f90 [hunk 4]
```diff
@@ -649,4 +669,122 @@ pure function replace_all_char_char_char(string, pattern, replacement) result(re
...
+        character(len=1), intent(in) :: pad_with
+        character(len=max(len(string), output_length)) :: res
+        integer :: string_length
+
+        string_length = len(string)
+
+        res = string
+        if (string_length < output_length) then
+            res(string_length + 1 : output_length) = repeat(pad_with, &
+            & output_length - string_length)
```
Comment (milancurcic) [inline src/stdlib_strings.f90:784]: I find the line break in the middle of function arguments list somewhat awkward, what do you think about breaking it at the assignment operator, i.e. LHS on one line and RHS on the other?

```suggestion
            res(string_length + 1 : output_length) = &
              repeat(pad_with, output_length - string_length)
```

I noticed that you use the continuation symbol `&` both at the end of the continuing line and at the beginning of the continued line. It's not necessary, you can have it only on the continuing line. But maybe you use both as a visual aid, which I think is fine.

**Review by milancurcic — APPROVED** [review]
Thank you @Aman-Godara, great addition, I left two comments.

**Comment by awvwgk** [comment]
Sorry for the merge conflict, I probably merged the PRs in the wrong order.

**Pull request merged**